x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineTests.java

```diff
@@ -531,4 +550,47 @@ public void close() throws IOException {
             }
         };
     }
+
+    public void testProcessOnceOnPrimary() throws Exception {
+        final Settings settings = Settings.builder().put("index.number_of_shards", 1).put("index.number_of_replicas", 0)
+            .put("index.version.created", Version.CURRENT).put("index.xpack.ccr.following_index", true).build();
+        final IndexMetaData indexMetaData = IndexMetaData.builder(index.getName()).settings(settings).build();
+        final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+        int numOps = between(10, 100);
+        List<Engine.Operation> operations = new ArrayList<>(numOps);
+        for (int i = 0; i < numOps; i++) {
+            ParsedDocument doc = EngineTestCase.createParsedDoc(Integer.toString(between(1, 100)), null);
+            if (randomBoolean()) {
+                operations.add(new Engine.Index(EngineTestCase.newUid(doc), doc, i, primaryTerm.get(), 1L,
+                    VersionType.EXTERNAL, Engine.Operation.Origin.PRIMARY, threadPool.relativeTimeInMillis(), -1, true));
+            } else {
+                operations.add(new Engine.Delete(doc.type(), doc.id(), EngineTestCase.newUid(doc), i, primaryTerm.get(), 1L,
+                    VersionType.EXTERNAL, Engine.Operation.Origin.PRIMARY, threadPool.relativeTimeInMillis()));
+            }
+        }
+        Randomness.shuffle(operations);
+        try (Store store = createStore(shardId, indexSettings, newDirectory())) {
+            final EngineConfig engineConfig = engineConfig(shardId, indexSettings, threadPool, store, logger, xContentRegistry());
+            try (FollowingEngine followingEngine = createEngine(store, engineConfig)) {
+                followingEngine.advanceMaxSeqNoOfUpdatesOrDeletes(operations.size() - 1L);
+                for (Engine.Operation op : operations) {
+                    Engine.Result result = applyOperation(followingEngine, op, randomFrom(Engine.Operation.Origin.values()));
+                    assertThat(result.getResultType(), equalTo(Engine.Result.Type.SUCCESS));
+                }
+                // Primary should reject duplicates
+                for (Engine.Operation op : operations) {
+                    Engine.Result result = applyOperation(followingEngine, op, Engine.Operation.Origin.PRIMARY);
+                    assertThat(result.getResultType(), equalTo(Engine.Result.Type.FAILURE));
+                    assertThat(result.getFailure(), instanceOf(AlreadyProcessedFollowingEngineException.class));
+                }
+                // Replica should accept duplicates
+                for (Engine.Operation op : operations) {
+                    Engine.Operation.Origin nonPrimary = randomValueOtherThan(Engine.Operation.Origin.PRIMARY,
+                        () -> randomFrom(Engine.Operation.Origin.values()));
+                    Engine.Result result = applyOperation(followingEngine, op, nonPrimary);
```

Comment:
any chance we can also check that this wasn't indexed to lucene? maybe doc counts?